src/signal.cr

```diff
@@ -90,67 +25,215 @@ require "c/unistd"
+  IO     = LibC::SIGIO
+  XCPU   = LibC::SIGXCPU
+  XFSZ   = LibC::SIGXFSZ
+  VTALRM = LibC::SIGVTALRM
+  USR1   = LibC::SIGUSR1
+  USR2   = LibC::SIGUSR2
+  WINCH  = LibC::SIGWINCH
+
+  {% if flag?(:linux) %}
+    PWR    = LibC::SIGPWR
+    STKFLT = LibC::SIGSTKFLT
+    UNUSED = LibC::SIGUNUSED
+  {% end %}
+
   # Sets the handler for this signal to the passed function.
   #
   # After executing this, whenever the current process receives the
   # corresponding signal, the passed function will be run (instead of the OS
   # default).
-  def trap(block : Signal ->)
-    trap &block
-  end
-
-  # ditto
-  def trap(&block : Signal ->)
-    Event::SignalHandler.add_handler self, block
+  def trap(&handler : Signal ->) : Nil
+    Crystal::SignalHandler.trap(self, handler)
   end
 
   # Resets the handler for this signal to the OS default.
-  def reset
-    case self
-    when CHLD
-      # don't ignore by default.  send events to a waitpid service
-      trap do
-        Event::SignalChildHandler.instance.trigger
-      end
-    else
-      del_handler Proc(Int32, Void).new(Pointer(Void).new(0_u64), Pointer(Void).null)
-    end
+  def reset : Nil
+    Crystal::SignalHandler.reset(self)
   end
 
   # Clears the handler for this signal and prevents the OS default action.
-  def ignore
-    del_handler Proc(Int32, Void).new(Pointer(Void).new(1_u64), Pointer(Void).null)
+  def ignore : Nil
+    Crystal::SignalHandler.ignore(self)
   end
 
-  private def del_handler(block)
-    Event::SignalHandler.del_handler self
-    LibC.signal value, block
-  end
-
-  @@default_handlers_setup = false
+  @@setup_default_handlers = Atomic(Int32).new(0)
 
   # :nodoc:
   def self.setup_default_handlers
-    return if @@default_handlers_setup
-    @@default_handlers_setup = true
+    _, success = @@setup_default_handlers.compare_and_set(0, 1)
+    return unless success
 
+    Crystal::SignalHandler.spawn
     Signal::PIPE.ignore
     Signal::CHLD.reset
   end
 end
 
 # :nodoc:
-fun __crystal_sigfault_handler(sig : LibC::Int, addr : Void*)
-  Crystal.restore_blocking_state
+module Crystal::SignalHandler
+  # The number of libc functions that can be called safely from a signal(2)
+  # handler is very limited. An usual safe solution is to use a pipe(2) and
+  # just write the signal to the file descriptor and nothing more. A loop in
+  # the main program is responsible for reading the signals back from the
+  # pipe(2) and handle the signal there.
+
+  @@pipe = IO.pipe(read_blocking: false, write_blocking: true)
+  @@procs = {} of Signal => Signal ->
+  @@mutex = Mutex.new
+
+  def self.trap(signal, handler) : Nil
+    @@mutex.synchronize do
+      unless @@procs[signal]?
+        LibC.signal(signal.value, ->(value : Int32) {
+          writer.write_bytes(value)
+        })
+      end
+      @@procs[signal] = handler
+    end
+  end
+
+  def self.reset(signal) : Nil
+    set(signal, Proc(Int32, Void).new(Pointer(Void).new(0_u64), Pointer(Void).null))
```

Comment:
How about `s/0_u64/LibC::SIG_DFL/`?

Reply:
We don't have them, sadly, and they're actually whole proc, not just an integer:
```c
#define SIG_DFL ((sighandler_t)0)
#define SIG_IGN ((sighandler_t)1)
```

Reply:
Passing in a proc with a fake function pointer just for the proc to be destroyed and only the function pointer used seems ugly too. Oh well, its not something that would prevent it from being merged.

Reply:
Done.